contributors/design-proposals/key-set-as-merge-key.md

```diff
@@ -0,0 +1,417 @@
+we will introduce additional merge keys in the recommended merge keys.
+- For the others, the recommended merge keys will be the same as the default merge key.
+
+Requirements for the recommended merge keys to keep backward compatibility:
+- the recommended merge keys must have the default merge key
+- the default merge key must be the first one in the recommended merge keys, i.e. `recommendedPatchMergeKey:"<default-merge-key>,<additional-key1>,<additional-key2>"`
+- the default merge key must be present in the patch, i.e. the first key in the recommended merge keys must be used.
+
+E.g. [`Ports` in `ServiceSpec`](https://github.com/kubernetes/kubernetes/blob/c51efa9ba0929a643544078d5c182ba75e4b4087/pkg/api/v1/types.go#L2825-L2831).
+```go
+type ServiceSpec struct {
+  // add recommendedPatchMergeKey "port,protocol"
+  Ports []ServicePort `patchMergeKey:"port" recommendedPatchMergeKey:"port,protocol" ...`
+  ...
+}
+```
+
+We also need to introduce a new optional directive `$patchMergeKey` in the patch.
+This directive contains a list of strings which are the merge keys used in this patch.
+
+An example patch will look like:
+```yaml
+list:
+- $patchMergeKey:
+  - foo
+  - bar
+  foo: a
+  bar: x
+  other: val
+```
+
+We will use the default merge key for merging if the $patchMergeKey directive is not present in the patch.
+Otherwise, we will use the merge keys provided by the directive.
+
+*Note*: Operations will take effect on all matching items in the list.
+
+#### When `$patchMergeKey` contains items that are not in the patch
+
+The item in the `$patchMergeKey` list but not in the patch will be considered as:
+when merging, this key is required to be not present to match an item in the list.
+
+This will be helpful when we want to distinguish the 2 items in the following list when merging:
+```yaml
+list:
+# foo is the default merge key; foo and bar are the recommended merge key.
+- foo: a
+- foo: a
+  bar: x
+```
+
+To match the first item, the patch should look like:
+```yaml
+list:
+- $patchMergeKey:
+  - foo
+  - bar
+  foo: a
+  # bar not present
+  other: val
+```
+
+To match the second item, the patch should look like:
+```yaml
+list:
+- $patchMergeKey:
+  - foo
+  - bar
+  foo: a
+  bar: x # bar presents
+  other: val
+```
+
+#### When no `$patchMergeKey` in the patch
+
+When there is no `$patchMergeKey` list in the patch,
+the server will use the default merge key when merging.
+
+E.g.
+foo is the default merge key.
+foo and bar are the recommended merge keys.
+
+Patch:
+```yaml
+list:
+- foo: a
+  bar: x
+  other: val
+```
+
+Live list:
+```yaml
+list:
+- foo: a
+  bar: y
+```
+
+Result after merging:
+```yaml
+list:
+- foo: a
+  bar: x
+  other: val
+```
+
+#### When `$patchMergeKey` only has part of all merge keys
+
+When `$patchMergeKey` only has part of all merge keys,
+the server will apply the patch to all the matching items.
+So an update or a deletion may apply to multiple items in the list.
```

Comment:
Accepting a `$patchMergeKey` that does not uniquely identify elements is what this proposal is trying to fix... I don't think a `$patchMergeKey` key that does not uniquely identify items should be accepted.

If it was, how would this interact with the proposed `$setElementOrder` directive, which is supposed to contain a list of keys? Keys that identify more than one item in the list mean that directive becomes ambiguous:

server items:
```yaml
ports:
- port: 8080
  protocol: TCP
  name: http-tcp
- port: 9090
  protocol: TCP
- port: 8080
  protocol: UDP
  name: http-udp
```

patch:
```yaml
$setElementOrder/ports:
- port: 8080
- port: 8080
- port: 9090
ports:
- port: 8080
  protocol: UDP
  name: http-udp-renamed
```

It's not clear which 8080 port was ordered where.